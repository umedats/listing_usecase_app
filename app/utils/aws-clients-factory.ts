import { EventBridge } from "@aws-sdk/client-eventbridge";
import { SQS } from "@aws-sdk/client-sqs";
import { SecretsManager } from "@aws-sdk/client-secrets-manager";

const AWS_CLIENTS = new Map<string, any>();
const EVENT_BRIDGE_SERVICE = "EventBridge";
const SQS_SERVICE = "SQS";
const SECRETS_MANAGER_SERVICE = "SecretsManager";

function computeCacheKey(service: string, region: string) {
  return `${service}-${region}`;
}

function buildAndCacheAWSClient(
  service: string,
  region: string,
  createClient: (region: string) => any,
) {
  const cacheKey = computeCacheKey(service, region);
  if (AWS_CLIENTS.has(cacheKey)) {
    return AWS_CLIENTS.get(cacheKey);
  } else {
    const client = createClient(region);
    AWS_CLIENTS.set(cacheKey, client);
    return client;
  }
}

/**
 * Helper method which builds an event bridge client.
 * @param region the AWS region.
 */
export function buildEventBridgeClient(region: string) {
  return buildAndCacheAWSClient(
    EVENT_BRIDGE_SERVICE,
    region,
    (region) =>
      new EventBridge({
        region: region,
      }),
  ) as EventBridge;
}

/**
 * Helper method which builds a sqs client.
 * @param region the AWS region.
 */
export function buildSQSClient(region: string) {
  return buildAndCacheAWSClient(
    SQS_SERVICE,
    region,
    (region) =>
      new SQS({
        region: region,
      }),
  ) as SQS;
}

/**
 * Helper method which builds a secrets manager client.
 * @param region the AWS region.
 */
export function buildSecretsManagerClient(region: string) {
  return buildAndCacheAWSClient(
    SECRETS_MANAGER_SERVICE,
    region,
    (region) =>
      new SecretsManager({
        region: region,
      }),
  ) as SecretsManager;
}
