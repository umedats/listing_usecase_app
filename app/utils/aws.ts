import { NOTIFICATION_TYPE_NORM_NAMES } from "@/app/constants/global";

/**
 * Helper method to generate the Event Bus Filter Rule name.
 * @param notificationType the notification type.
 */
export function buildEventBusFilterRuleName(notificationType: string) {
  const notificationNormName =
    NOTIFICATION_TYPE_NORM_NAMES.get(notificationType) ?? "";
  return `${notificationNormName}Rule`;
}

/**
 * Helper method to generate the SQS Queue name.
 * @param notificationType the notification type.
 */
export function buildNotificationSQSQueueName(notificationType: string) {
  const notificationNormName =
    NOTIFICATION_TYPE_NORM_NAMES.get(notificationType) ?? "";
  return `${notificationNormName}Queue`;
}

/**
 * Helper method to generate the SQS Arn for the notification Type.
 *
 * @param region Region of the SQS queue.
 * @param awsAccountId AWS account where the queue will be created.
 * @param notificationType Notification type for which the SQS Queue ARN will be generated.
 */
export function buildNotificationSQSQueueArn(
  region: string,
  awsAccountId: string,
  notificationType: string,
) {
  return `arn:aws:sqs:${region}:${awsAccountId}:${buildNotificationSQSQueueName(
    notificationType,
  )}`;
}

/**
 * Helper method to generate the SQS target name.
 * @param notificationType the notification type.
 */
export function buildNotificationSQSTargetName(notificationType: string) {
  const notificationNormName =
    NOTIFICATION_TYPE_NORM_NAMES.get(notificationType) ?? "";
  return `${notificationNormName}Target`;
}

/**
 * Helper method which generates the SQS Queue Url.
 * @param awsAccountId the AWS Account Id.
 * @param region the AWS region.
 * @param queueName the SQS queue name.
 */
export function generateSQSQueueUrl(
  awsAccountId: string,
  region: string,
  queueName: string,
) {
  return `https://sqs.${region}.amazonaws.com/${awsAccountId}/${queueName}`;
}
