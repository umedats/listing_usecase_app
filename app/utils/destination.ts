import { SPAPIRequestResponse } from "@/app/model/types";
import getSPAPIEndpoint, {
  buildGrantLessNotificationsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  CREATE_DESTINATION_API_DOC_LINK,
  CREATE_DESTINATION_API_NAME,
  EVENT_BRIDGE_DESTINATION_NAME,
  GET_DESTINATIONS_API_DOC_LINK,
  GET_DESTINATIONS_API_NAME,
} from "@/app/constants/global";
import {
  CreateDestinationRequest,
  Destination,
  DestinationList,
  GetDestinationsResponse,
} from "@/app/sdk/notifications";

/**
 * Helper method which invokes the GetDestinations API to fetch all the
 * event bridge destinations.
 * @param region the AWS region.
 * @param awsAccountId the AWS awsAccountId.
 * @param reqResponseCollector collects all the SP API requests and responses.
 */
export async function getEventBridgeDestinations(
  region: string,
  awsAccountId: string,
  reqResponseCollector: SPAPIRequestResponse[],
) {
  const notificationsApi = await buildGrantLessNotificationsAPIClient(
    getSPAPIEndpoint(region),
  );
  const requestResponse: SPAPIRequestResponse = {
    apiName: GET_DESTINATIONS_API_NAME,
    apiDocumentationLink: GET_DESTINATIONS_API_DOC_LINK,
    request: {},
    response: {},
  };

  try {
    const getDestinationsResponse: GetDestinationsResponse =
      GetDestinationsResponse.constructFromObject(
        await notificationsApi.getDestinations(),
        undefined,
      );
    requestResponse.response = getDestinationsResponse;
    const destinations: Destination[] = DestinationList.constructFromObject(
      getDestinationsResponse.payload,
      undefined,
    );

    return destinations?.filter(
      (destination) =>
        destination?.resource?.eventBridge?.accountId === awsAccountId &&
        destination?.resource?.eventBridge?.region === region,
    );
  } catch (error: any) {
    requestResponse.response = error;
    if (!(error?.status === 404)) {
      throw error;
    }
  } finally {
    reqResponseCollector.push(requestResponse);
  }
}

/**
 * Helper method which invokes the CreateDestination API to create new event bridge destination.
 * @param awsAccountId the AWS awsAccountId.
 * @param region the AWS region.
 * @param reqResponseCollector collects all the SP API requests and responses.
 */
export async function createEventBridgeDestination(
  awsAccountId: string,
  region: string,
  reqResponseCollector: SPAPIRequestResponse[],
) {
  const notificationsApi = await buildGrantLessNotificationsAPIClient(
    getSPAPIEndpoint(region),
  );
  const destinationResourceSpecification = {
    eventBridge: {
      region: region,
      accountId: awsAccountId,
    },
  };
  const createDestinationRequestBody =
    CreateDestinationRequest.constructFromObject(
      {
        resourceSpecification: destinationResourceSpecification,
        name: EVENT_BRIDGE_DESTINATION_NAME,
      },
      undefined,
    );
  const requestResponse = {
    apiName: CREATE_DESTINATION_API_NAME,
    apiDocumentationLink: CREATE_DESTINATION_API_DOC_LINK,
    request: {
      body: createDestinationRequestBody,
    },
    response: {},
  };

  try {
    requestResponse.response = await notificationsApi.createDestination(
      requestResponse.request.body,
    );
  } catch (error: any) {
    requestResponse.response = error;
    throw error;
  } finally {
    reqResponseCollector.push(requestResponse);
  }
  return requestResponse.response;
}
