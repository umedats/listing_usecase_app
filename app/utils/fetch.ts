import { SPAPIRequestResponse } from "@/app/model/types";
import { SettableDebugState } from "@/app/context/debug-context-provider";
import { NO_CACHING_DIRECTIVE } from "@/app/constants/global";

/**
 * Helper function to fetch the data from the backend API. This function
 * extracts the debug context, and invokes the updateDebugContextCallback
 * method.
 * @param url the url of the backend API.
 * @param updateDebugContextCallback a callback to update the debug context.
 * @param responseDataCallback an optional call back which is invoked with the
 * data extracted from the backend API.
 * @param requestInit params to initialize the fetch request.
 */
export async function fetcherForBackendAPI(
  url: string,
  updateDebugContextCallback: (debugData: any) => void,
  responseDataCallback?: (data: any) => any | void,
  requestInit?: RequestInit,
) {
  let newRequestInit = requestInit;
  if (requestInit) {
    newRequestInit = { ...requestInit, cache: NO_CACHING_DIRECTIVE };
  }

  const response = await fetch(url, newRequestInit);
  const responseBody = await response.json();
  if (responseBody.debugContext) {
    updateDebugContextCallback(responseBody.debugContext);
  }

  if (!response.ok) {
    throw [response.status, response.statusText];
  }

  const data = responseBody.data;
  if (responseDataCallback) {
    return responseDataCallback(data);
  }
  return data;
}

/**
 * Saves the given debugData for the given route into debugContext.
 * @param pathname
 * @param debugData
 * @param debugContext
 */
export function updateDebugContext(
  pathname: string,
  debugData: any,
  debugContext: SettableDebugState,
) {
  (debugData as SPAPIRequestResponse[]).forEach((value) =>
    debugContext.addRequestResponse(pathname, value),
  );
}
