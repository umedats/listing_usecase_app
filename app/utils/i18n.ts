/**
 * Helper function which converts the locales like en-US to en_US.
 * @param bcp47Locale
 */
export function convertLocaleToSPAPIFormat(bcp47Locale: string) {
  const intlLocale = new Intl.Locale(bcp47Locale);
  return `${intlLocale.language}_${intlLocale.region}`;
}
