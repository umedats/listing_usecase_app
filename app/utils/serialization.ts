/**
 * A helper method to serialize a given object into Json string.
 * @param data any data which needs to be serialized.
 */
export function serializeToJsonString(data: any) {
  if (typeof data === "bigint") {
    return (data as BigInt).toString(10);
  }

  try {
    return JSON.stringify(data, null, 4);
  } catch (error) {
    console.log("Unable to serialize the given data due to error : " + error);
    return "";
  }
}
