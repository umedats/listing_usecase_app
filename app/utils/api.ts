import nextResponse from "@/app/utils/next-response-factory";
import { getSettings, SettingsResponse } from "@/app/api/settings/wrapper";
import { serializeToJsonString } from "@/app/utils/serialization";
import { STATUS } from "@/app/constants/global";
import { SPAPIRequestResponse } from "@/app/model/types";
import { Settings } from "@/app/[locale]/settings/settings";
import { NextRequest, NextResponse } from "next/server";

/**
 * Generic Error Message which is sent in the API response.
 */
const ERROR_MESSAGE = "Unexpected error during API execution.";

/**
 * Helper function to build 500 HTTP response on settings fetch failure.
 */
export async function retrieveSettingsData() {
  const settingsResponse: SettingsResponse = await getSettings();

  if (settingsResponse.status !== 200) {
    const errorMessage =
      "Unable to retrieve the settings due to " + settingsResponse.statusText;
    console.log(errorMessage);
    return nextResponse(500, errorMessage);
  }

  return settingsResponse.settings;
}

/**
 * Helper method which retrieves the settings and invokes the API logic.
 * @param apiLogic the function which executes the api logic.
 */
export async function retrieveSettingsAndInvokeAPILogic(
  apiLogic: (settings: Settings) => Promise<NextResponse>,
) {
  try {
    const settingsResponse = await retrieveSettingsData();

    if (settingsResponse instanceof NextResponse) {
      return settingsResponse;
    }

    const settings: Settings = settingsResponse;
    return apiLogic(settings);
  } catch (error) {
    return nextResponse(500, ERROR_MESSAGE, serializeToJsonString({}));
  }
}

/**
 * Helper method which retrieves the settings and invokes the API logic with next request.
 * @param apiLogic the function which executes the api logic.
 * @param apiRequest the next request.
 */
export async function retrieveSettingsAndInvokeAPILogicWithRequest(
  apiLogic: (
    settings: Settings,
    apiRequest: NextRequest,
  ) => Promise<NextResponse>,
  apiRequest: NextRequest,
) {
  try {
    const settingsResponse = await retrieveSettingsData();

    if (settingsResponse instanceof NextResponse) {
      return settingsResponse;
    }

    const settings: Settings = settingsResponse;
    return apiLogic(settings, apiRequest);
  } catch (error) {
    return nextResponse(500, ERROR_MESSAGE, serializeToJsonString({}));
  }
}

/**
 * Helper function to handle any error from the SP API.
 * @param reqResponse the object to store the SP API request and response.
 * @param error SP API error.
 */
export function handleSPAPIInvocationError(
  reqResponse: SPAPIRequestResponse,
  error: any,
) {
  reqResponse.response = error;
  // The error is generally the Error object from the superagent.
  const errorMessage = ERROR_MESSAGE;
  console.log(errorMessage + " " + serializeToJsonString(error));

  // Proxy the status from the SP API.
  let status = 500;
  if (error instanceof Object && error.hasOwnProperty(STATUS)) {
    status = error.status;
  }

  return nextResponse(
    status,
    errorMessage,
    serializeToJsonString({
      debugContext: [reqResponse],
    }),
  );
}

/**
 * Helper function to handle any error from the SP API where multiple SP APIs are chained within a single request.
 * @param error SP API error.
 * @param debugContext List where debug context will be appended. Use the same list for all calls.
 * @param reqResponse the object to store the SP API request and response.
 */
export function handleChainedSPAPIInvocationError(
  error: any,
  debugContext: any[],
  reqResponse?: SPAPIRequestResponse,
) {
  if (reqResponse) {
    reqResponse.response = error;
    debugContext.push(reqResponse);
  }

  // The error is generally the Error object from the superagent.
  const errorMessage = ERROR_MESSAGE;
  console.log(errorMessage + " " + serializeToJsonString(error));

  // Proxy the status from the SP API.
  let status = 500;
  if (error instanceof Object && error.hasOwnProperty(STATUS)) {
    status = error.status;
  }

  return nextResponse(
    status,
    errorMessage,
    serializeToJsonString({
      debugContext: debugContext,
    }),
  );
}
