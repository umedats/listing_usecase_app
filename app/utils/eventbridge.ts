import {
  EventBridge,
  ListRulesCommandOutput,
  ListTargetsByRuleCommandOutput,
  RemoveTargetsCommandOutput,
  ResourceAlreadyExistsException,
  ResourceNotFoundException,
  Rule,
  RuleState,
  Target,
} from "@aws-sdk/client-eventbridge";
import { serializeToJsonString } from "@/app/utils/serialization";
import { buildEventBridgeClient } from "@/app/utils/aws-clients-factory";
import {
  buildEventBusFilterRuleName,
  buildNotificationSQSQueueArn,
  buildNotificationSQSTargetName,
} from "@/app/utils/aws";
import { createQueue } from "@/app/utils/sqs";

function handleEventBridgeError(error: any) {
  console.log(serializeToJsonString(error));
  if (!(error instanceof ResourceNotFoundException)) {
    throw {
      status: 500,
    };
  }
}

/**
 * Helper method which returns the [Rule, Target] information for the given
 * eventbus.
 * @param region the aws region.
 * @param eventBusName the event bus name.
 */
export async function getRuleTargetsForEventBus(
  region: string,
  eventBusName: string,
) {
  const ruleTargets: { rule: Rule; target: Target }[] = [];
  const eventBridge = buildEventBridgeClient(region);
  try {
    let listRulesNextToken = undefined;
    do {
      const listRulesCommandOutput: ListRulesCommandOutput =
        await eventBridge.listRules({
          EventBusName: eventBusName,
          NextToken: listRulesNextToken,
        });
      listRulesNextToken = listRulesCommandOutput?.NextToken;

      const rules = listRulesCommandOutput?.Rules ?? [];
      for (const rule of rules) {
        let listTargetsByRuleNextToken = undefined;
        do {
          const listTargetsByRuleCommandOutput: ListTargetsByRuleCommandOutput =
            await eventBridge.listTargetsByRule({
              EventBusName: eventBusName,
              Rule: rule.Name,
            });
          listTargetsByRuleNextToken =
            listTargetsByRuleCommandOutput?.NextToken;
          const targets = listTargetsByRuleCommandOutput?.Targets ?? [];
          for (const target of targets) {
            ruleTargets.push({
              rule: rule,
              target: target,
            });
          }
        } while (listTargetsByRuleNextToken);
      }
    } while (listRulesNextToken);
  } catch (error) {
    handleEventBridgeError(error);
  }
  return ruleTargets;
}

async function removeTarget(
  eventBridge: EventBridge,
  eventBusName: string,
  rule: string,
  targetId: string,
) {
  try {
    const removeTargetsCommandOutput: RemoveTargetsCommandOutput =
      await eventBridge.removeTargets({
        EventBusName: eventBusName,
        Rule: rule,
        Ids: [targetId],
      });

    if (removeTargetsCommandOutput?.FailedEntryCount !== 0) {
      removeTargetsCommandOutput.FailedEntries?.forEach((failedEntry) => {
        console.log(serializeToJsonString(failedEntry));
      });
      handleEventBridgeError("removeTargets API returned 1 FailedEntryCount");
    }
  } catch (error) {
    handleEventBridgeError(error);
  }
}

async function createTarget(
  eventBridge: EventBridge,
  eventBusName: string,
  region: string,
  awsAccountId: string,
  notificationType: string,
) {
  try {
    const putTargetsResponse = await eventBridge.putTargets({
      EventBusName: eventBusName,
      Rule: buildEventBusFilterRuleName(notificationType),
      Targets: [
        {
          Arn: buildNotificationSQSQueueArn(
            region,
            awsAccountId,
            notificationType,
          ),
          Id: buildNotificationSQSTargetName(notificationType),
        },
      ],
    });

    if (putTargetsResponse?.FailedEntryCount !== 0) {
      putTargetsResponse.FailedEntries?.forEach((failedEntry) => {
        console.log(serializeToJsonString(failedEntry));
      });
      handleEventBridgeError(
        "putTargets API returned at least 1 FailedEntryCount",
      );
    }
  } catch (error: any) {
    if (!(error instanceof ResourceAlreadyExistsException)) {
      handleEventBridgeError(error);
    }
  }
}

async function deleteRule(
  eventBridge: EventBridge,
  eventBusName: string,
  rule: string,
) {
  try {
    await eventBridge.deleteRule({
      EventBusName: eventBusName,
      Name: rule,
    });
  } catch (error) {
    handleEventBridgeError(error);
  }
}

async function createRule(
  eventBridge: EventBridge,
  eventBusName: string,
  notificationType: string,
) {
  try {
    await eventBridge.putRule({
      Name: buildEventBusFilterRuleName(notificationType),
      EventBusName: eventBusName,
      Description: getEventBusRuleDescription(notificationType),
      EventPattern: getSPAPIEventPatternForEventBusRule(notificationType),
      State: RuleState.ENABLED,
    });
  } catch (error: any) {
    if (!(error instanceof ResourceAlreadyExistsException)) {
      handleEventBridgeError(error);
    }
  }
}

/**
 * Helper method which removes the target and deletes the rule.
 * @param eventBusName name of the event bus.
 * @param rule the event bus rule to delete.
 * @param targetId the Id of the target to remove from the rule.
 * @param region the AWS region.
 */
export async function removeTargetAndDeleteRule(
  eventBusName: string,
  rule: string,
  targetId: string,
  region: string,
) {
  const eventBridge = buildEventBridgeClient(region);
  await removeTarget(eventBridge, eventBusName, rule, targetId);
  await deleteRule(eventBridge, eventBusName, rule);
}

async function createAndAssociateEventBusHelper(
  eventBridge: EventBridge,
  eventBusName: string,
) {
  try {
    await eventBridge.createEventBus({
      Name: eventBusName,
      EventSourceName: eventBusName,
    });
  } catch (error: any) {
    if (!(error instanceof ResourceAlreadyExistsException)) {
      handleEventBridgeError(error);
    }
  }
}

/**
 * Helper method to create(If one does not exist already) and associate Event Bus.
 *
 * @param eventBusName EventBus Name.
 * @param region AWS region.
 */
export async function createAndAssociateEventBus(
  eventBusName: string,
  region: string,
) {
  const eventBridge = buildEventBridgeClient(region);
  await createAndAssociateEventBusHelper(eventBridge, eventBusName);
}

/**
 * Helper method to create the necessary Event Bridge resources
 * 1. Create New Event Bus Rule.
 * 2. Create New SQS Queue(For each notification type, if not created already).
 * 3. Create New Target.
 *
 * @param eventBusName EventBus Name.
 * @param notificationType Type of the Notification.
 * @param awsAccountId AWS Account associated with the application.
 * @param region AWS region.
 */
export async function createEventBridgeResources(
  eventBusName: string,
  notificationType: string,
  awsAccountId: string,
  region: string,
) {
  const eventBridge = buildEventBridgeClient(region);
  await createRule(eventBridge, eventBusName, notificationType);
  await createQueue(notificationType, region, awsAccountId, eventBusName);
  await createTarget(
    eventBridge,
    eventBusName,
    region,
    awsAccountId,
    notificationType,
  );
}

function getEventBusRuleDescription(notificationType: string) {
  return `Event Bus rule to receive the ${notificationType} notifications from the selling partner API.`;
}

// Function to get the SP-API Event Pattern for EventBus Rule.
// https://tinyurl.com/bdf98t9f
function getSPAPIEventPatternForEventBusRule(notificationType: string) {
  return `
    {
      "source": [
        {
          "prefix": "aws.partner/sellingpartnerapi.amazon.com"
        }
      ],
      "detail-type": ["${notificationType}"]
    }
  `;
}
