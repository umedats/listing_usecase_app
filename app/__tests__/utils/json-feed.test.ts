import {
  computePatches,
  convertListingToJsonFeed,
} from "@/app/utils/json-feed";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  CREATE_LISTING_USE_CASE,
  CREATE_OFFER_USE_CASE,
  PATCH_LISTINGS_ITEM_API_NAME,
  UPDATE_LISTING_USE_CASE,
} from "@/app/constants/global";

const SKU: string = "SKU";
const PRODUCT_TYPE: string = "SHIRT";
const SELLER_ID: string = "ATVPDKIKX0DER";
const ISSUE_LOCALE: string = "en_US";
const CURRENT_LISTING = {
  item_name: [
    {
      value: "Title",
      marketplace_id: "ATVPDKIKX0DER",
    },
  ],
};
const EMPTY_INITIAL_LISTING = {};

describe("Test for the Json Feed", () => {
  test("verify the converted feed content for CreateOffer use case", async () => {
    const feed = convertListingToJsonFeed(
      SKU,
      PRODUCT_TYPE,
      CURRENT_LISTING,
      CREATE_OFFER_USE_CASE,
      EMPTY_INITIAL_LISTING,
      SELLER_ID,
      ISSUE_LOCALE,
    );
    expect(serializeToJsonString(feed)).toMatchSnapshot();
  });

  test("verify the converted feed content for CreateListing use case ", async () => {
    const feed = convertListingToJsonFeed(
      SKU,
      PRODUCT_TYPE,
      CURRENT_LISTING,
      CREATE_LISTING_USE_CASE,
      EMPTY_INITIAL_LISTING,
      SELLER_ID,
      ISSUE_LOCALE,
    );
    expect(serializeToJsonString(feed)).toMatchSnapshot();
  });

  test("verify the converted feed content for UpdateListing use case ", async () => {
    const feed = convertListingToJsonFeed(
      SKU,
      PRODUCT_TYPE,
      {
        keyAdded: [
          {
            value: "keyAdded",
          },
        ],
        keyUnChanged: [
          {
            value: "keyUnChanged1",
            keywords: ["Keyword1", "KeyWord2"],
          },
          {
            value: "keyUnChanged2",
            keywords: ["Keyword3", "KeyWord4"],
          },
        ],
        keyChanged: [
          {
            value: "keyUnChanged",
          },
        ],
      },

      UPDATE_LISTING_USE_CASE,
      {
        keyRemoved: [
          {
            value: "keyRemoved",
          },
        ],
        keyUnChanged: [
          {
            value: "keyUnChanged1",
            keywords: ["Keyword1", "KeyWord2"],
          },
          {
            value: "keyUnChanged2",
            keywords: ["Keyword3", "KeyWord4"],
          },
        ],
        keyChanged: [
          {
            value: "keyUnChanged1",
          },
        ],
      },
      SELLER_ID,
      ISSUE_LOCALE,
      PATCH_LISTINGS_ITEM_API_NAME,
    );
    expect(serializeToJsonString(feed)).toMatchSnapshot();
  });

  test("returned undefined on empty attributes for CreateOffer usecase", () => {
    const feed = convertListingToJsonFeed(
      SKU,
      PRODUCT_TYPE,
      {},
      "CreateOfferListing",
      {},
      SELLER_ID,
      ISSUE_LOCALE,
    );
    expect(feed).toBeUndefined();
  });

  test("returned undefined on empty patches for UpdateListing usecase", () => {
    const listing = {
      item_name: [
        {
          value: "Title",
        },
      ],
    };
    const feed = convertListingToJsonFeed(
      SKU,
      PRODUCT_TYPE,
      listing,
      "CreateOfferListing",
      listing,
      SELLER_ID,
      ISSUE_LOCALE,
      PATCH_LISTINGS_ITEM_API_NAME,
    );
    expect(feed).toBeUndefined();
  });

  test("computePatches creates a patch for the first common attribute", () => {
    const listing = {
      item_name: [
        {
          value: "Title",
        },
      ],
    };
    const initialListing = {
      item_name: [
        {
          value: "Title1",
        },
      ],
    };
    const patches = computePatches(listing, initialListing);
    expect(patches?.length).toStrictEqual(1);
    expect(patches[0]).toMatchSnapshot();
  });
});
