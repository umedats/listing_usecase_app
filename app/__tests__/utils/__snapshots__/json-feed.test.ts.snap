// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the Json Feed computePatches creates a patch for the first common attribute 1`] = `
{
  "op": "replace",
  "path": "/attributes/item_name",
  "value": [
    {
      "value": "Title",
    },
  ],
}
`;

exports[`Test for the Json Feed verify the converted feed content for CreateListing use case  1`] = `
"{
    "header": {
        "sellerId": "ATVPDKIKX0DER",
        "version": "2.0",
        "issueLocale": "en_US"
    },
    "messages": [
        {
            "messageId": 1,
            "sku": "SKU",
            "operationType": "UPDATE",
            "productType": "SHIRT",
            "requirements": "LISTING",
            "attributes": {
                "item_name": [
                    {
                        "value": "Title",
                        "marketplace_id": "ATVPDKIKX0DER"
                    }
                ]
            }
        }
    ]
}"
`;

exports[`Test for the Json Feed verify the converted feed content for CreateOffer use case 1`] = `
"{
    "header": {
        "sellerId": "ATVPDKIKX0DER",
        "version": "2.0",
        "issueLocale": "en_US"
    },
    "messages": [
        {
            "messageId": 1,
            "sku": "SKU",
            "operationType": "UPDATE",
            "productType": "SHIRT",
            "requirements": "LISTING_OFFER_ONLY",
            "attributes": {
                "item_name": [
                    {
                        "value": "Title",
                        "marketplace_id": "ATVPDKIKX0DER"
                    }
                ]
            }
        }
    ]
}"
`;

exports[`Test for the Json Feed verify the converted feed content for UpdateListing use case  1`] = `
"{
    "header": {
        "sellerId": "ATVPDKIKX0DER",
        "version": "2.0",
        "issueLocale": "en_US"
    },
    "messages": [
        {
            "messageId": 1,
            "sku": "SKU",
            "operationType": "PATCH",
            "productType": "SHIRT",
            "patches": [
                {
                    "op": "add",
                    "path": "/attributes/keyAdded",
                    "value": [
                        {
                            "value": "keyAdded"
                        }
                    ]
                },
                {
                    "op": "delete",
                    "path": "/attributes/keyRemoved",
                    "value": [
                        {
                            "value": "keyRemoved"
                        }
                    ]
                },
                {
                    "op": "replace",
                    "path": "/attributes/keyChanged",
                    "value": [
                        {
                            "value": "keyUnChanged"
                        }
                    ]
                }
            ]
        }
    ]
}"
`;
